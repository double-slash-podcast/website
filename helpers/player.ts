export type typeDuration = {minutes: number; seconds: number};

/**
 * get duration for media file
 * @param length: number
 * @returns {{seconds: number, minutes: number}}
 */
export const calculateTotalValue = (length: number): typeDuration => {
  const minutes = Math.floor(length / 60);
  const seconds = +(length - minutes * 60).toFixed(1);
  return {minutes, seconds};
};
