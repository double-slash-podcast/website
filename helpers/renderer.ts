import {h, VNode} from 'vue';
import {toHtml} from 'hast-util-to-html';

interface NodeBase {
  tagName?: string;
  properties?: {[key: string]: string};
  type: string;
  tag?: string;
  value?: string;
  props?: {[key: string]: string};
}

interface NodeType extends NodeBase {
  children: NodeType[];
}

const renderMDToHtml = (node: NodeType): VNode => {
  if (node.type === 'text') return h(node);
  if (node.tag) {
    return h(node.tag, node.props, node.children?.map(renderMDToHtml));
  }
  if (node.type === 'root' && node.children) {
    return node.children?.map(e => renderMDToHtml(e));
  }
  return h('div');
};

const convertToAstFormat = node => {
  if (node.type === 'text') return node;
  if (node.type === 'root' && node.children) {
    return node.children.map(convertToAstFormat);
  }
  node.tagName = node.tag;
  node.properties = node.props;
  node.children = node.children.map(convertToAstFormat);
  return node;
};

export const convertToHtml = (node: NodeType): string =>
  toHtml(convertToAstFormat(node));

export default renderMDToHtml;
