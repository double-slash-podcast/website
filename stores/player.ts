import {defineStore} from 'pinia';

export const usePlayerStore = defineStore('player', () => {
  const config = useRuntimeConfig();
  const {prefixAudioDev, prefixAudio, isDev} = config.public;

  // source for player
  const src = ref<string | undefined>(undefined);
  // current title of podcast player
  const currentTitle = ref<string | undefined>(undefined);
  // current status of player
  const statusPlayer = ref<typeStatusPlayer>('pause');

  /**
   * update source player and title
   */
  function setDsSlug(dsSlug: string | undefined, title: string): void {
    if (!dsSlug) {
      throw new Error(`dsSlug is undefined for ${title}`);
    }
    // remove tracking for dev
    src.value = `${isDev ? prefixAudioDev : prefixAudio}/${dsSlug}.mp3`;
    currentTitle.value = title;
  }

  /**
   * update status player
   */
  function setStatusPlayer(status: typeStatusPlayer) {
    statusPlayer.value = status;
  }

  return {src, currentTitle, statusPlayer, setDsSlug, setStatusPlayer};
});
