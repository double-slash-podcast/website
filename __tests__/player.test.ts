import {describe, expect, test} from 'vitest';
import {calculateTotalValue} from '../helpers/player';

describe('test duration detail', () => {
  test('get object with detailed duration 1:08:55', () => {
    const t = calculateTotalValue(4135);
    expect(t).toMatchSnapshot();
  });

  test('get object with detailed duration 0:51:10', () => {
    const t = calculateTotalValue(3070);
    expect(t).toMatchSnapshot();
  });
});
