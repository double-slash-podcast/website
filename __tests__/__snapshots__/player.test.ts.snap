// Vitest Snapshot v1

exports[`test duration detail > get object with detailed duration 0:51:10 1`] = `
{
  "hours": 0,
  "minutes": 51,
  "seconds": 10,
}
`;

exports[`test duration detail > get object with detailed duration 1:08:55 1`] = `
{
  "hours": 1,
  "minutes": 8,
  "seconds": 55,
}
`;
